### PR TITLE
Add some functions to inspect the raw LSeq entries as opposed to just their values

### PR DESCRIPTION
**Pull request by bochaco**

### CODE REVIEW

**Comment by bochaco** [comment]
@davidrusu I'm working on supporting Access Control on top of an LSEQ, and for such thing I need access to the raw `Entry` to be able to figure out some causality and also for undoing operations in some scenarios on a replica (at the moment by copying items onto a fresh LSEQ filtering by `Identifier`). I'm hoping these functions which I'm needing are generic enough for having them in here as part of the LSEQ API.

File: src/lseq.rs [hunk 1]
```diff
@@ -223,16 +223,31 @@ impl<T: Clone, A: Actor> LSeq<T, A> {
...
+        self.seq.iter()
+    }
+
     /// Get an element at an index from the sequence represented by the LSEQ.
     pub fn get(&self, ix: usize) -> Option<&T> {
         self.seq.get(ix).map(|Entry { val, .. }| val)
     }
 
+    /// Finds an element searching by its Identifier.
+    pub fn find(&self, ident: &Identifier<A>) -> Option<&Entry<T, A>> {
```
Comment (davidrusu) [inline src/lseq.rs:237]: rename to `find_entry`?

File: src/lseq.rs [hunk 2]
```diff
@@ -223,16 +223,31 @@ impl<T: Clone, A: Actor> LSeq<T, A> {
...
+        self.seq.iter().find(|Entry { id, .. }| id == ident)
+    }
+
     /// Get last element of the sequence represented by the LSEQ.
     pub fn last(&self) -> Option<&T> {
         self.seq.last().map(|Entry { val, .. }| val)
```
Comment (davidrusu) [inline src/lseq.rs:243]: rewrite this to use `self.last_entry().map(...)`

File: src/lseq.rs [hunk 3]
```diff
@@ -223,16 +223,31 @@ impl<T: Clone, A: Actor> LSeq<T, A> {
...
+    pub fn iter_entries(&self) -> impl Iterator<Item = &Entry<T, A>> + '_ {
+        self.seq.iter()
+    }
+
     /// Get an element at an index from the sequence represented by the LSEQ.
     pub fn get(&self, ix: usize) -> Option<&T> {
         self.seq.get(ix).map(|Entry { val, .. }| val)
     }
 
+    /// Finds an element searching by its Identifier.
```
Comment (davidrusu) [inline src/lseq.rs:236]:
```suggestion
    /// Finds an entry searching by its Identifier.
```

**Comment by davidrusu** [comment]
> to figure out some causality and also for undoing operations on a replica in some scenarios

Very interesting, is this access control code up somewhere?

**Comment by bochaco** [comment]
I just pushed a commit with the naming changes.

> Very interesting, is this access control code up somewhere?

Yes, for now I have a very simple implementation where I'm just cloning the LSEQ for each new version of the Policy and I sometimes need to undo ops, or apply them on more than one copy of the LSEQ if the op came late and after a new Policy op: https://github.com/maidsafe/safe-nd/blob/ebd6318d2612dbf1b1fce8fced6d5ca8bae949a9/src/sequence/seq_crdt.rs.

You may wanna look at the tests I've created which show the scenarios I'm covering: https://github.com/maidsafe/safe-nd/blob/ebd6318d2612dbf1b1fce8fced6d5ca8bae949a9/src/sequence/mod.rs#L470

I was also trying with another approach, aiming at not copying the LSEQ but that was leading me to some impl complexities to cover scenarios of concurrent Policy ops. Another approach I was playing with would queue non-causally ready operations but we are trying to avoid such things like queuing/caching things: https://github.com/bochaco/safe-nd/blob/9e4c55a44233c9c61080e66b3fca7947c0a62c45/src/sequence/seq_crdt.rs

**Comment by davidrusu** [comment]
Interesting, since you only expose the append operation, then you can get away with just filtering out the identifier you want to undo. Am I right to think that this would get more complex if users were to be allowed to remove identifiers from your datatype?

**Comment by bochaco** [comment]
Yes, that's correct, that's the reason I put that comment trying to explain that is an "undo". I even thought about working on an 'undo()' function for the LSEQ API but it seems we'd need to keep an ops log which again it's something we are trying to avoid (GC is not trivial either). I haven't thought much more than that about this.

**Pull request merged**